crates/circuit/src/circuit_drawer.rs

```diff
@@ -0,0 +1,916 @@
+                        label: Some(label.clone()),
+                        visual_type: VisualType::Boxed,
+                    }
+                }
+            }
+        } 
+        else {
+            // print all operation details
+            println!("Unsupported operation details:");
+            //get operator discriminant
+            // let PackedOperation(discriminant) = instruction.op;
+            // println!("{064b}", discriminant);
+            println!("Name: {}", instruction.op.name());
+            println!("Parameters: {:?}", instruction.params_view());
+            println!("Qubits: {:?}", self.dag_circ.qargs_interner().get(instruction.qubits));
+            println!("Clbits: {:?}", self.dag_circ.cargs_interner().get(instruction.clbits));
+            panic!("Unsupported operation: {}", instruction.op.name());
+
+        }
+    }
+
+    pub fn new(dag_circ: DAGCircuit) -> Self {
+
+        //number of qubits in dag_circuit
+        let qubit = dag_circ.num_qubits();
+
+        circuit_rep {
+            q_wires: vec!(wire::new(Wire_Type::qubit); qubit as usize),
+            dag_circ: dag_circ
+        }
+    }
+
+    pub fn circuit_string(&self) -> String {
+        let mut output = String::new();
+        for wires in self.q_wires.iter() {
+            output.push_str(&wires.get_wire_rep());
+        }
+        output
+    }
+
+    pub fn fix_len(&mut self) {
+        let mut num = 0;
+        for wire in self.q_wires.iter_mut() {
+            let length = wire.get_len();
+            if length > num {
+                num = length;
+            }
+        }
+
+        for wire in self.q_wires.iter_mut() {
+            let length = wire.get_len();
+            wire.fix_len(num - length);
+        }
+    }
+
+    pub fn set_qubit_name(&mut self) {
+        for (i, qubit) in self.dag_circ.qubits().objects().iter().enumerate() {
+            let qubit_name = if let Some(locations) = self.dag_circ.qubit_locations().get(qubit) {
+                if let Some((register, reg_index)) = locations.registers().first() {
+                    format!("{}_{}", register.name(), reg_index)
+                } else {
+                    format!("q_{}", i)
+                }
+            } else {
+                format!("q_{}", i)
+            };
+            self.q_wires[i].qubit_name(&qubit_name);
+        }
+        self.fix_len();
+    }
+
+    pub fn build_layer(&mut self, layer: Vec<&PackedInstruction>){
+        for instruction in layer{
+            let components = {
+                let enclosed = self.from_instruction(instruction);
+                enclosed.component_dict()
+            };
+            println!("components: {:?}", components);
+            for (idx, component) in components{
+                self.q_wires[idx as usize].add_wire_component(&component);
+            }
+        }
+        // normalise all wire widths
+        // self.fix_len();
+    }
+
+    pub fn build_layers(&mut self) {
```

Comment:
It's better, i.e. more modular, if this function just returns the layers to the caller.